h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamMojoModelBase.java

```diff
@@ -0,0 +1,171 @@
+package hex.genmodel.algos.gam;
+
+import hex.genmodel.ConverterFactoryProvidingModel;
+import hex.genmodel.GenModel;
+import hex.genmodel.MojoModel;
+import hex.genmodel.easy.CategoricalEncoder;
+import hex.genmodel.easy.EasyPredictModelWrapper;
+import hex.genmodel.easy.RowData;
+import hex.genmodel.easy.RowToRawDataConverter;
+import hex.genmodel.utils.ArrayUtils;
+import hex.genmodel.utils.DistributionFamily;
+import hex.genmodel.utils.LinkFunctionType;
+
+import java.util.Map;
+import java.util.Set;
+
+public abstract class GamMojoModelBase extends MojoModel implements ConverterFactoryProvidingModel {
+  String _link;
+  public LinkFunctionType _link_function;
+  boolean _useAllFactorLevels;
+  int _cats;
+  int[] _catNAFills;
+  int[] _catOffsets;
+  int _nums;
+  int _numsCenter;
+  double[] _numNAFills;
+  double[] _numNAFillsCenter;
+  boolean _meanImputation;
+  double[] _beta;
+  double[] _beta_no_center;
+  double[] _beta_center;
+  double[][] _beta_multinomial;
+  double[][] _beta_multinomial_no_center; // coefficients not centered for multinomial/ordinal
+  double[][] _beta_multinomial_center; // coefficients not centered for multinomial/ordinal
+  DistributionFamily _family;
+  String[] _gam_columns;
+  int _num_gam_columns;
+  int[] _bs;
+  int[] _num_knots;
+  double[][] _knots;
+  double[][][] _binvD;
+  double[][][] _zTranspose;
+  String[][] _gamColNames;  // expanded gam column names
+  String[][] _gamColNamesCenter;
+  String[] _names_no_centering; // column names of features with no centering
+  int _totFeatureSize; // number of predictors plus gam columns no centered
+  int _betaSizePerClass;
+  int _betaCenterSizePerClass;
+  double _tweedieLinkPower;
+  double[][] _basisVals; // store basis values array for each gam column
+  double[][] _hj;   // difference between knot values
+  int _numExpandedGamCols; // number of expanded gam columns
+  int _lastClass;
+  int _predictColumnNum=0; // Count number of columns that are found both in input data row and gam predictor columns
+  
+  public int getTotFeatureSize() { return _totFeatureSize;}
+  public String[] getNames() { return _names; }  // get column names centered
+  public int getPredictColumnNum() { return _predictColumnNum;}
+  
+  GamMojoModelBase(String[] columns, String[][] domains, String responseColumn) {
+    super(columns, domains, responseColumn);
+  }
+
+  public void setPredictColumNum(Set<String> inputColumnNames) {
+     for (String cName : features()) {
+      if (inputColumnNames.contains(cName))
+        _predictColumnNum++; // count column names that are present both in input data row and gam predictor names
```

Comment:
I still feel like this logic is unnecessary, in addExpandGamCols you expect the exact size of the row to be features+gam_cols, in now way has _predictColumnNum impact on that method

Reply:
@honza:  When you send in the training dataset, all you have is predictors, columns chosen to be transformed to the gam_columns,  ignored columns and response column.  Then, gam will add expanded gam_cols to the training dataset and the features will include just the predictors and the expanded gam_cols and the response column.

Then, you are given the test dataset which should include the predictors, columns chosen to be transformed and other columns that we don't care about like the ignored columns and others.  The setPredictColumNum is just trying to find out how many predictor columns the test dataset contain when compared to the features that the model desired.  

I thought about getting rid of it too cannot find a way around it.